opendcs-rest-api/src/main/java/org/opendcs/odcsapi/res/DataSourceResources.java

```diff
@@ -52,78 +57,231 @@ public class DataSourceResources
+		}
+	}
+
+	static ArrayList<ApiDataSourceRef> map(DataSourceList dsl)
+	{
+		ArrayList<ApiDataSourceRef> ret = new ArrayList<>();
+		for(DataSource ds : dsl.getList())
+		{
+			ApiDataSourceRef adr = new ApiDataSourceRef();
+			if (ds.getId() != null)
+			{
+				adr.setDataSourceId(ds.getId().getValue());
+			}
+			else
+			{
+				adr.setDataSourceId(DbKey.NullKey.getValue());
+			}
+			adr.setName(ds.getName());
+			adr.setType(ds.dataSourceType);
+			adr.setUsedBy(ds.numUsedBy);
+			adr.setArguments(map(ds.getArguments()));
+			ret.add(adr);
+		}
+		return ret;
+	}
+
+	static String map(Properties props)
+	{
+		if (props == null || props.isEmpty())
+			return null;
+
+		StringBuilder retVal = new StringBuilder();
+		for (Object key : props.keySet())
+		{
+			retVal.append(key).append("=").append(props.getProperty((String) key)).append(",");
+		}
+		return retVal.toString();
 	}
 
 	@GET
 	@Path("datasource")
 	@Produces(MediaType.APPLICATION_JSON)
 	@RolesAllowed({AuthorizationCheck.ODCS_API_GUEST})
 	public Response geDataSource(@QueryParam("datasourceid") Long dataSourceId)
-		throws WebAppException, DbException, SQLException
+			throws WebAppException, DbException
 	{
 		if (dataSourceId == null)
-			throw new WebAppException(ErrorCodes.MISSING_ID, 
-				"Missing required datasourceid parameter.");
-		
-		Logger.getLogger(ApiConstants.loggerName).fine("getDataSource id=" + dataSourceId);
-		try (DbInterface dbi = new DbInterface();
-			ApiDataSourceDAO dao = new ApiDataSourceDAO(dbi))
-		{
-			ApiDataSource ret = dao.readDataSource(dataSourceId);
-			if (ret == null)
-				throw new WebAppException(ErrorCodes.NO_SUCH_OBJECT, 
-					"No such DECODES data source with id=" + dataSourceId + ".");
-			return ApiHttpUtil.createResponse(ret);
+		{
+			throw new WebAppException(ErrorCodes.MISSING_ID,
+					"Missing required datasourceid parameter.");
 		}
+
+		try
+		{
+			DatabaseIO dbio = getLegacyDatabase();
+			DataSource ds = new DataSource(DbKey.createDbKey(dataSourceId));
+			dbio.readDataSource(ds);
+
+			if (ds.getName() == null)
+			{
+				throw new WebAppException(ErrorCodes.NO_SUCH_OBJECT,
+						"No such DECODES data source with id=" + dataSourceId + ".");
+			}
+			ApiDataSource ret = map(ds);
+			return Response.status(HttpServletResponse.SC_OK).entity(ret).build();
+		}
+		catch (DatabaseException ex)
+		{
+			throw new DbException("Error reading data source: " + ex);
+		}
+	}
+
+	static ApiDataSource map(DataSource ds)
+	{
+		if (ds == null)
+			return null;
+		ApiDataSource ads = new ApiDataSource();
+		if (ds.getId() != null)
+		{
+			ads.setDataSourceId(ds.getId().getValue());
+		}
+		else
+		{
+			ads.setDataSourceId(DbKey.NullKey.getValue());
+		}
+		ads.setName(ds.getName());
+		ads.setType(ds.dataSourceType);
+		ads.setProps(ds.getArguments());
+		ads.setGroupMembers(map(ds.groupMembers));
+		return ads;
+	}
+
+	static ArrayList<ApiDataSourceGroupMember> map(Vector<DataSource> groupMembers)
```

Comment:
@MikeNeilson is there a reason the DTO classes have to use ArrayList and Vector?  Can we just fix the DTO so that we don't have to return implementation classes?

Reply:
Not really. In a few cases it's because the Original DAOs or Interfaces used Vector and they ArrayList. So there would need to be a conversion step but I would argue we should just do so.

The DAOs are going to be moved to List or Collection as the output, but we haven't gotten to it yet.